### ISSUE TITLE
Support `import.meta.env`

### ISSUE THREAD

**Issue by JounQin** [webpack/webpack#15833]
## Feature request

**What is the expected behavior?**

`import.meta.env.NODE_ENV`

**What is motivation or use case for adding/changing the behavior?**

migrate `process.env.NODE_ENV`

**How should this be implemented in your opinion?**

No idea

**Are you willing to work on this yourself?**

No idea

**Comment by vankop** [comment]
which tools support it? only Vite.js?

**Comment by philippevezina** [comment]
> which tools support it? only Vite.js?

To my knowledge yes.

Webpack 5 has taken a big step by stepping away from Node in bundled code by removing NodeJS library polyfills from being included in bundles. Removing `process` seems like the next big step in the same direction.

ViteJS also provides a powerful tool to replace CommonJS `require` statements: [`import.meta.glob` and `import.meta.globEager`](https://vitejs.dev/guide/features.html#glob-import).

**Comment by vankop** [comment]
yeah, webpack supports `import.meta.webpackContext` https://github.com/webpack/webpack/pull/15446

**Comment by alexander-akait** [comment]
For example we can found many issues like https://github.com/immerjs/immer/issues/557 (and workaround https://github.com/immerjs/immer/issues/557#issuecomment-621491287) for these packages, `import.meta.env` is an unofficial solution for any platform.

**Comment by alexander-akait** [comment]
Anyway we can merge https://github.com/webpack/webpack/pull/15841/ to solve problems with packages which already migrate on `import.meta.url`, but the question is more bigger, will be great if developers from Node.js, Deno, Rollup, Parcel, Esbuild and other had an discussion here. It is not good that vite just decide to use `import.meta.env` for env and start to recommend it, it can be destructive for ecosystem.

**Comment by binyamin** [comment]
@alexander-akait Just FYI, I looked at the tc39 proposal for `import.meta`. It seems the ECMAScript committee decided it _should_ be extensible.[^1] Particular hosts, such as Node.js and Deno, may lock down the API if they choose. I wouldn't extend it for a Node.js script. But in the browser, it's fair game.

[^1]: https://github.com/tc39/proposal-import-meta#will-this-object-be-locked-down-in-any-way

**Comment by JounQin** [comment]
ping

**Comment by alexander-akait** [comment]
bump, still need a dicussion

**Comment by XiaoLin1995** [comment]
You can try to use [import-meta-loader](https://www.npmjs.com/package/import-meta-loader).
It will support you to use vite's import.meta in webpack. Because it can convert vite's import.meta syntax to webpack's syntax.

example:
```js
// vite => webpack
import.meta.env //=> process.env
import.meta.env.MODE //=> process.env.NODE_ENV
import.meta.env.BASE_URL //=> process.env.BASE_URL
import.meta.env.PROD //=> process.env.NODE_ENV === 'production'
import.meta.env.DEV //=> process.env.NODE_ENV !== 'production'

new URL('filePath', import.meta.url).href //=> require('filePath')

import.meta.glob('filePath') //=> ... require.context('dirPath', useSubdirectories: boolean, RegExp, 'lazy') ...
import.meta.glob('filePath', { eager: true }) //=> vite3 ... require.context('dirPath', useSubdirectories: boolean, RegExp, 'sync') ...
import.meta.globEager('filePath') //=> vite2 ... require.context('dirPath', useSubdirectories: boolean, RegExp, 'sync') ...
```

**Comment by JounQin** [comment]
I'm like to have a more standard way because `process` is not standard global variable.

**Comment by silverwind** [comment]
The modern way of doing it in node is an explicit import:

```js
import {env} from 'node:process';
console.info(env.NODE_ENV);
```

Equvialent for webpack, maybe the runtime code of webpack could export it? Are there existing interfaces where user code interfaces with webpack code?

```js
import {env} from 'webpack:runtime';
console.info(env.NODE_ENV);
```

**Comment by saolof** [comment]
I'd rather encourage the explicit import as mentioned above than rely on the bundler monkey patching attributes global variables that are then used in the code, after just having been bit by the problem. 

The dependency on the .env attributes leads to code that _fails at runtime_ if you build it with the wrong bundler. This is just technical debt imposed on anyone that uses a bundler with this kind of API in their projects.

**Comment by ravshansbox** [comment]
@saolof it’s not a matter of how to include values, but rather generate an artifact with the values from the desired environment. One can’t do it with normal imports.

**Comment by benmccann** [comment]
The problem with the explicit import is that if you're writing a library that you want to be usable on multiple platforms you have to import a value for each different platform and check them all. It's much nicer to have some standardization. If webpack and Vite agree how to do it that is the majority of the market and others can follow
